src/python/WMCore/Services/DDM/DDM.py

```diff
@@ -0,0 +1,286 @@
+        msg += "API: {}: ddmReq: {}".format(ddmReq.api, pformat(ddmReq))
+        self['logger'].info(msg)
+
+        # The Request class expects this to be a dict, otherwise it tries to
+        # encode the inputdata once again and inserts parasite information and
+        # breaks the call to dynamo.
+        ddmReq = dict(ddmReq)
+        data = self.refreshCache(cachedApi,
+                                 apiUrl,
+                                 inputdata=ddmReq,
+                                 verb='POST')
+        results = data.read()
+        data.close()
+
+        results = json.loads(results)
+        return results
+
+    def makeRequest(self, ddmReq, api=None):
+        """
+        _makeRequest_
+
+        Makes a SINGLE request to DDM
+
+        :ddmReq: Request template an instance of DDMReqTemplate class.
+        :api:    The DDM api to be queried.
+        :return: Dictionary containing the result from DDM
+        """
+
+        # try to figure out the api from the ddmReq template
+        if not api:
+            api = ddmReq.api
+
+        if not api:
+            msg = "WARNING: skip sending data to DDM: "
+            msg += "API: {}: ddmReq: {}\n".format(ddmReq.api, pformat(ddmReq))
+            return None
+
+        apiUrl = "registry/request/{}".format(api)
+        callname = "DDM-{}".format(api)
+
+        result = None
+        try:
+            # TODO: here/getResult to send the request parameters to the logger
+            result = self._getResult(ddmReq, apiUrl, callname)
+        except Exception as e:
+            msg = "ERROR: sending data to DDM: \n"
+            msg += "API: {}\n".format(ddmReq.api)
+            msg += "ddmReq: {}\n".format(pformat(ddmReq))
+            msg += "ERROR: {}".format(str(e))
+            self['logger'].debug(msg)
+            return None
+        return result
+
+    def makeAggRequests(self, ddmReqList, aggKey=None):
+        """
+        _makeAggRequest_
+
+        Makes several Aggregated requests to DDM. Takes the list of objects,
+        instances of ddmReqTemplate and aggregates it based on a selected key.
+        It iterates through the list and compares all to all. Uses the
+        ddmReqTemplate internal Equivalence method for comparison with the
+        aggregation key excluded. If a match is found the two requests are
+        squeezed into one. The final result should be a new list of unique
+        ddmRequests each one of them should contain all similar ddmRequests
+        from the initial list with the values of the aggregation key accumulated
+        in a list under the same key name. Creates a request per ddmReqTemplate
+        from the list with aggregated templates.
+        NOTE:
+            In this method the api is always fetched from the template.
+        TODO:
+            So far it is working basically for the 'copy' API and due to that
+            the aggregation key is always a dictionary so no explicit check is
```

Comment:
"the aggregation key is always a dictionary" makes be believe that `aggKey` is a dictionary, and not a string as stated below.

Reply:
Yes, This comment is to stress exactly this fact. That it is working so far only for keys, which have dictionary values (and as we already checked for 'item' DDM does accept dictionary). If we decide to make it functional for other aggkeys we should have an iteration with DDM people to ask which are the keys in a request that can be lists and which cannot.

Reply:
Done